text/trace/0168-sampling-propagation.md

```diff
@@ -0,0 +1,419 @@
+
+To develop an intuition for r-values, consider a scenario where every
+bit of the `TraceID` is generated by a uniform random bit generator
+(i.e., every bit is 0 or 1 with equal probability).  An 128-bit
+`TraceID` can therefore be treated as a 128-bit unsigned integer,
+which can be mapped into a fraction with range [0, 1) by dividing by
+2**128, a form known as the TraceID-ratio.  Now, probability sampling
+could be achieved by comparing the TraceID-ratio with the sampling
+probability, setting the `sampled` flag when TraceID-ratio is less
+than the sampling probability.
+
+It is easy to see that with sampling probability 1, all TraceIDs will
+be accepted because TraceID ratios are exclusively less than 1.
+Sampling with probability 50% will select TraceID ratios less than
+0.5, which maps to all TraceIDs less than 2**127 or, equivalently, all
+TraceIDs where the most significant bit is zero.  By the same logic,
+sampling with probability 25% means accepting TraceIDs where the most
+significant two bits are zero.  In general, with exact probability
+`2**-S` is equivalent to selecting TraceIDs with `S` leading zeros in
+this example scenario.
+
+The r-value specified here directly describes the number of leading
+zeros in a random 61-bit string, specified in a way that does not
+require TraceID values to be constructed with random bits in specific
+positions or with hard requirements on their uniformity.  In
+mathematical terms, the r-value is described by a truncated geometric
+distribution, listed below:
+
+| `r` Value        | Probability of `r-value` | Implied sampling probabilities |
+| ---------------- | ------------------------ | ----------------------         |
+| 0                | 1/2                      | 1                              |
+| 1                | 1/4                      | 1/2 and above                  |
+| 2                | 1/8                      | 1/4 and above                  |
+| 3                | 1/16                     | 1/8 and above                  |
+| ...              | ...                      | ...                            |
+| 0 <= `r` <= 60   | 1/(2**(-`r`-1))          | 2**-`r` and above              |
+| ...              | ...                      | ...                            |
+| 58               | 2**-59                   | 2**-58 and above               |
+| 59               | 2**-60                   | 2**-59 and above               |
+| 60               | 2**-61                   | 2**-60 and above               |
+| 61               | 2**-61                   | 2**-61 and above               |
+
+Such a random variable `r` can be generated using efficient
+instructions on modern computer architectures, for example:
+
+```golang
+import (
+	"math/rand"
+	"math/bits"
+)
+
+func nextRValueLeading() int {
+	x := uint64(rand.Int63()) // 63 least-significant bits are random
+	y := x << 1 | 0x7         // 61 most-significant bits are random
+	return bits.LeadingZeros64(y)
+}
```

Comment:
alternative which is language & std-lib neutral (may be off by one):
```go
x := uint64(rand.Int63())
for r := 0; r < 61; r++ {
    if x & 0x1 == 0x1 {
        return r
    }
}
return 61
```

Reply:
I added this as a second example, computing the number of trailing zeros. I added it to the gist and verified that it computes the same result as the version using `bits.TrailingZeros64()`.